bibtex-actions.el

```diff
@@ -234,6 +234,36 @@ If FORCE-REBUILD-CACHE is t, force reloading the cache."
     (bibtex-actions-refresh force-rebuild-cache))
   bibtex-actions--candidates-cache)
 
+(defun bibtex-actions-complete-key-at-point ()
+  "Complate citation key at point.
+
+When inserting '@' in a buffer the capf UI will present user with
+a list of entries, from which they can narrow against title,
+author, etc., and then select one. This function will then return
+the key 'key', resulting in '@key' at point."
+
+  ;; Activate when user enters '@'.
+  ;;
+  ;; Notes:
+  ;;
+  ;;   - pandoc syntax is [@key], or can also be just @key.
+  ;;   - org syntax is stricter; [cite:see @key1 p23]
+  ;;   - latex uses keys, but sans '@' (I don't want to worry about this ATM though)
+  ;;
+  ;; Maybe this should be constrained further for org?
+  ;;
+  ;; Or, given that pandoc syntax can be used in org context, maybe condition
+  ;; this on 'bibtex-completion-format-citation-functions'?
+  (when (string-equal "@" (string (preceding-char)))
+    ;; Use same candidates as 'bibtex-read'.
+    (let (candidates (bibtex-actions--get-candidates))
+      (list
+       ;; TODO I adapted this from @minad's completion-in-region example, but
+       ;; not sure if it's correct.
+       (- (point) 1) (point)
+       ;; TODO need to return the cdr of the candidate; not the car.
+       candidates :exclusive 'no))))
```

Comment:
Could you please help me with this @minad?

The candidate car is the search string, and the cdr is the key I need to insert.

Reply:
@bdarcus What is the question here exactly? 

In case you are looking for an example capf, you can try this:

~~~ elisp
(defun test-capf ()
  (list
   (save-excursion
     (backward-word)
     (point))
   (point)
   (lambda (str pred action)
     (message "CALLED %S %S" str action)
     (complete-with-action action '("a1" "a2" "a3" "a4"
                                    "b1" "b2" "b3" "b4"
                                    "c1" "c2" "c3" "c4")
                           str pred))))

(add-to-list 'completion-at-point-functions #'test-capf)
~~~

Reply:
Btw. the PR code is unreadable here on the gh web, due this gh action noise. I cannot see the wood for the trees.

Reply:
> Btw. the PR code is unreadable here on the gh web, due this gh action noise

Sorry about that. Fixed.

I'll take a look at your example. Thanks.

Reply:
@minad 

> In case you are looking for an example capf, you can try this:

I finally got back to this, and managed to get your example working with corfu!

So if I understand right, I probably need a new function to create the completion table, that I can call from both this function and from the completing-read one. Correct?

And how do I get it to activate with `@`, and for the result at point to be `@key` (as it would be in org-mode and with pandoc; latex does not include the `@`, but I'll worry about that later)?

The completion table has keys, without the `@`.

Reply:
OK, I adapted your example, @minad, to mirror my case. 

The completion UI now works as I want ...

![Screenshot from 2021-06-02 10-50-30](https://user-images.githubusercontent.com/1134/120502558-86404d00-c390-11eb-9b24-83d95dda51f1.png)

... but I'm stuck on one last thing: how do I get the cdr of the candidate instead of the car?

``` elisp
(defun test-capf ()
  "My capf."
  (when (looking-back "@[a-zA-Z]+")
    (list
     (save-excursion
       (backward-word)
       (point))
     (point)
     (lambda (str pred action)
       (let ((candidates '(("ay one" . "a1")
                           ("ay two" . "a2")
                           ("ay three" . "a3")
                           ("ay four" . "a4"))))
       (message "CALLED %S %S" str action)
       (complete-with-action action candidates str pred))))))
```